Making it possible to pass Windows credential specs directly to the engine

wk8 opened the pull request:
Instead of having to go through files or registry values as is currently the
case.

While adding GMSA support to Kubernetes (https://github.com/kubernetes/kubernetes/pull/73726)
I stumbled upon the fact that Docker currently only allows passing Windows
credential specs through files or registry values, forcing the Kubelet
to perform a rather awkward dance of writing-then-deleting to either the
disk or the registry to be able to create a Windows container with cred
specs.

This patch solves this problem by making it possible to directly pass
whole base64-encoded cred specs to the engine's API. I took the opportunity
to slightly refactor the method responsible for Windows cred spec as it
seemed hard to read to me.

Added some unit tests on Windows credential specs handling, as there were
previously none.

Added/amended the relevant integration tests.

I have also tested it manually: given a Windows container using a cred spec
that you would normally start with e.g.
```powershell
docker run --rm --security-opt "credentialspec=file://win.json" mcr.microsoft.com/windows/servercore:ltsc2019 nltest /parentdomain
# output:
# my.ad.domain.com. (1)
# The command completed successfully
```
can now equivalently be started with
```powershell
$rawCredSpec = & cat 'C:\ProgramData\docker\credentialspecs\win.json'
$escaped = $rawCredSpec.Replace('"', '\"')
docker run --rm --security-opt "credentialspec=raw://$escaped" mcr.microsoft.com/windows/servercore:ltsc2019 nltest /parentdomain
# same output!
```

I'll do another PR on Swarmkit after this is merged to allow services to use
the same option.

(It's worth noting that @dperny faced the same problem adding GMSA support
to Swarmkit, to which he came up with an interesting solution - see
https://github.com/moby/moby/pull/38632 - but alas these tricks are not
available to the Kubelet.)

thaJeztah commented on daemon/oci_windows.go:
```diff
@@ -344,6 +285,82 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
+			if credentialSpec, err = readCredentialSpecRegistry(c.ID, value); err != nil {
+				return err
+			}
+		case "config":
```
wondering (haven't fully thought this through); should we remove the `config` option, and make swarmkit convert it to `base64` (read the credential-spec from the config, and pass it as base64)?

@dperny @cpuguy83

wk8 replied:
I had the same question, actually :)

cpuguy83 replied:
Yes, please ❤️

wk8 replied:
Happy to make a PR for that on Swarmkit after this is merged is @dperny likes the idea too :)

cpuguy83 replied:
Does it require a swarmkit change? I think the node agent can just read the config and stick it into the the credential spec?

wk8 replied:
@cpuguy83 : you may very well be right, I didn't look into it before opening my big mouth, much to my shame. I just assumed that it did require a swarmkit change.

wk8 replied:
Would still very much like @dperny 's take on this though before I do make that change :)

dperny replied:
It would theoretically be a bit of a pain in the butt, but we could do that. You'd basically have to read the CredentialSpec at some point when converting to a container, and write it to base64. It should be done all in the engine, no swarmkit changes necessary.

thaJeztah replied:
@wk8 just thinking of this; the credentials-spec file is always JSON, correct? Do we need to pass it as base64, or should we just pass as a plain string? (optionally, validating for well-formedness)

wk8 replied:
Yes it is always JSON, so it could just be passed as a plain string.

In my experience though, that can fast become a pain with quote escaping when eg building a command line in a bash script, that's why I thought base64 might be a better idea? Happy to change though if you feel strongly about it.

JSON validation is an idea though. Didn't do it upfront because we don't seem to validate anything just yet - if we do validate, we should do it for all the different ways that exist; and that would arguably be somewhat of a breaking change. Happy to add it if you wish though.

thaJeztah replied:
> In my experience though, that can fast become a pain with quote escaping when eg building a command line in a bash script, that's why I thought base64 might be a better idea

You mean when using (e.g.) `curl` to create API calls? Or when using a bash script that's using the (`docker`) cli?

If we want to add this as an option on the docker cli, I assume the cli itself would take care of properly encoding the JSON. (perhaps the option should be named `raw` in that case)

Should `stdin` be supported there? e.g.:

```
cat ./my-creds.json | docker service create --credential-specs=file://- ....
```

(using `-` as name for the file to indicate it's from `stdin`)

wk8 replied:
Fair enough, changing to using raw strings.

cpuguy83 replied:
Can we go ahead and remove config from this?

wk8 replied:
As we discussed on the maintainers' call yesterday, I'll do that in a follow-up PR.

cpuguy83 replied:
Why not just in a separate commit?

cpuguy83 replied:
Seems like the whole reason to add this is to remove config.

wk8 replied:
Well, the original reason for this is to remove the whole awkward dance I had to do in the kubelet to write/clean up registry keys to add GMSA support there (see the PR message).

And I'll be happy to do that in a follow-up PR; it can't hurt to keep PRs to a reasonable size?

cpuguy83 replied:
Fine, I would prefer to have the change come in through a single PR but in separate commits but 🤷‍♂️

wk8 replied:
It arguably is 2 different things; I'll make a follow-up PR as soon as this one is merged. Thanks!

thaJeztah commented on daemon/oci_windows.go:
```diff
@@ -471,7 +491,7 @@ func readCredentialSpecFile(id, root, location string) (string, error) {
 	bcontents, err := ioutil.ReadFile(full)
 	if err != nil {
-		return "", fmt.Errorf("credential spec '%s' for container %s as the file could not be read: %q", full, id, err)
+		return "", fmt.Errorf("credential spec for container %s could not be read from file %q: %q", id, full, err)
```
I saw @dperny was working on some of those, but while we're changing this, we should change these to;

```suggestion
		return "", errors.Wrapf(err, "credential spec for container %s could not be read from file %q", id, full)
```

wk8 replied:
Good call, doing :)

thaJeztah commented on daemon/oci_windows.go:
```diff
@@ -427,34 +444,37 @@ func (daemon *Daemon) mergeUlimits(c *containertypes.HostConfig) {
+			return "", fmt.Errorf("registry credential spec %q for container %s was not found", name, id)
 		}
-		return "", fmt.Errorf("error %v reading credential spec %q from registry for container %s", err, name, id)
+		return "", fmt.Errorf("error reading credential spec %q from registry for container %s: %v", name, id, err)
```
Same here

```suggestion
		return "", errors.Wrapf(err, "error reading credential spec %q from registry for container %s", name, id)
```

wk8 replied:
👍

thaJeztah commented on daemon/oci_windows.go:
```diff
@@ -427,34 +444,37 @@ func (daemon *Daemon) mergeUlimits(c *containertypes.HostConfig) {
-	if val, _, err = k.GetStringValue(name); err != nil {
+	key, err := registryOpenKeyFunc(registry.LOCAL_MACHINE, credentialSpecRegistryLocation, registry.QUERY_VALUE)
+	if err != nil {
+		return "", fmt.Errorf("failed handling spec %q for container %s - registry key %s could not be opened: %v", name, id, credentialSpecRegistryLocation, err)
```
```suggestion
		return "", errors.Wrapf(err, "failed handling spec %q for container %s - registry key %s could not be opened", name, id, credentialSpecRegistryLocation)
```

wk8 replied:
👍

dperny commented on daemon/oci_windows.go:
```diff
@@ -344,6 +285,82 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
 	return nil
 }
 
+var errInvalidCredentialSpecSecOpt = fmt.Errorf("invalid credential spec security option - value must be prefixed by 'file://', 'registry://', or 'base64://' followed by a non-empty value")
```
I have a PR in the works (#38770) to normalize all of the error values to use the `errdefs` definitions. Can you modify this and any other _new_ errors to use those error types, so that I don't have to repeat the work when rebasing that PR?

wk8 replied:
Sure thing, doing that today or tomorrow.

wk8 replied:
Done :)

thaJeztah commented:
Looks like there's failures on Windows;

```
01:50:37 FAIL: docker_cli_run_test.go:4146: DockerSuite.TestRunCredentialSpecFailures
01:50:37 
01:50:37 docker_cli_run_test.go:4159:
01:50:37     c.Assert(err.Error(), checker.Contains, attempt.expectedError, check.Commentf("%s expected %s got %s", attempt.value, attempt.expectedError, err))
01:50:37 ... obtained string = "" +
01:50:37 ...     "\n" +
01:50:37 ...     "Command:  d:\\CI\\CI-34c625c5ff\\binary\\docker.exe run --security-opt=credentialspec=rubbish busybox true\n" +
01:50:37 ...     "ExitCode: 125\n" +
01:50:37 ...     "Error:    exit status 125\n" +
01:50:37 ...     "Stdout:   \n" +
01:50:37 ...     "Stderr:   d:\\CI\\CI-34c625c5ff\\binary\\docker.exe: error during connect: Post http://127.0.0.1:2357/v1.30/containers/7e999460d094c24e4ebea2e71c3f8e33cdfd700ccfe3a476766cb6886ceff284/start: EOF.\n" +
01:50:37 ...     "\n" +
01:50:37 ...     "\n" +
01:50:37 ...     "Failures:\n" +
01:50:37 ...     "ExitCode was 125 expected 0\n" +
01:50:37 ...     "Expected no error"
01:50:37 ... substring string = "invalid credential spec security option - value must be prefixed file:// or registry://"
01:50:37 ... rubbish expected invalid credential spec security option - value must be prefixed file:// or registry:// got 
01:50:37 Command:  d:\CI\CI-34c625c5ff\binary\docker.exe run --security-opt=credentialspec=rubbish busybox true
01:50:37 ExitCode: 125
01:50:37 Error:    exit status 125
01:50:37 Stdout:   
01:50:37 Stderr:   d:\CI\CI-34c625c5ff\binary\docker.exe: error during connect: Post http://127.0.0.1:2357/v1.30/containers/7e999460d094c24e4ebea2e71c3f8e33cdfd700ccfe3a476766cb6886ceff284/start: EOF.
01:50:37 
01:50:37 
01:50:37 Failures:
01:50:37 ExitCode was 125 expected 0
01:50:37 Expected no error
```

wk8 commented:
@thaJeztah : thanks, having a look at the failure shortly.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "wk8/raw_cred_specs" git@github.com:wk8/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354465880
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

wk8 committed:
Making it possible to pass Windows credential specs directly to the engine

Instead of having to go through files or registry values as is currently the
case.

While adding GMSA support to Kubernetes (https://github.com/kubernetes/kubernetes/pull/73726)
I stumbled upon the fact that Docker currently only allows passing Windows
credential specs through files or registry values, forcing the Kubelet
to perform a rather awkward dance of writing-then-deleting to either the
disk or the registry to be able to create a Windows container with cred
specs.

This patch solves this problem by making it possible to directly pass
whole base64-encoded cred specs to the engine's API. I took the opportunity
to slightly refactor the method responsible for Windows cred spec as it
seemed hard to read to me.

Added some unit tests on Windows credential specs handling, as there were
previously none.

Added/amended the relevant integration tests.

I have also tested it manually: given a Windows container using a cred spec
that you would normally start with e.g.
```powershell
docker run --rm --security-opt "credentialspec=file://win.json" mcr.microsoft.com/windows/servercore:ltsc2019 nltest /parentdomain
# output:
# my.ad.domain.com. (1)
# The command completed successfully
```
can now equivalently be started with
```powershell
$rawCredSpec = & cat 'C:\ProgramData\docker\credentialspecs\win.json'
$escaped = $rawCredSpec.Replace('"', '\"')
docker run --rm --security-opt "credentialspec=raw://$escaped" mcr.microsoft.com/windows/servercore:ltsc2019 nltest /parentdomain
# same output!
```

I'll do another PR on Swarmkit after this is merged to allow services to use
the same option.

(It's worth noting that @dperny faced the same problem adding GMSA support
to Swarmkit, to which he came up with an interesting solution - see
https://github.com/moby/moby/pull/38632 - but alas these tricks are not
available to the Kubelet.)

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 commented:
@thaJeztah : changed to accept raw strings, and fixed the failing test (and added more). PTAL.

cpuguy83 reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM; left a suggestion for using `gotest.tools` for creating the directory and test-file, but let me know if you like that or not 😅

wk8 commented:
@thaJeztah : thanks, made the change you requested. Didn't change the way the file is written though, because that would have pushed the file creation too far from where it's actually used, and made the test more confusing IMO.

wk8 commented on daemon/oci_windows_test.go:
```diff
@@ -0,0 +1,314 @@
+
+func TestSetWindowsCredentialSpecInSpec(t *testing.T) {
+	// we need a temp directory to act as the daemon's root
+	tmpDaemonRoot, err := ioutil.TempDir("", "moby-windows-cred-spec-test-")
```
Can't do it for the second part, as I do need a specific dir name in there - and moving it all up to the very beginning of the test, far from where it's actually used, makes it a bit confusing IMO.

thaJeztah reviewed (approved):
LGTM, thanks!

wk8 commented:
Thanks! :)

Drewster727 commented:
@thaJeztah @wk8 Is there any documentation on this? 
I have a swarm setup with windows nodes and would like to pass credential specs. However, I do not want to add the spec file to each node, which is why I liked @wk8's original thought of passing it with the **service create** and compose.yml directly.

Is this something we can do now on 19.03?

Thanks,
Drew
